Bottom: 0306c4a8f653d9713685909e385da5b3df2339dd
Top:    cfef74d9fd3a3f86339c0927a71c75b0a19cd54f
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2008-10-12 03:12:36 +0100

Refresh of speedup_poly_poly_checking

---

diff --git a/src/find.c b/src/find.c
index a1b7167..e8d08c5 100644
--- a/src/find.c
+++ b/src/find.c
@@ -1896,6 +1896,30 @@ LOCtoLineRat_callback (const BoxType * b, void *cl)
 }
 
 static int
+LOCtoLinePolygon_callback (const BoxType * b, void *cl)
+{
+  PolygonTypePtr polygon = (PolygonTypePtr) b;
+  struct lo_info *i = (struct lo_info *) cl;
+
+  if (!TEST_FLAG (TheFlag, polygon) &&
+      IsLineInPolygon (&i->line, polygon) &&
+      ADD_POLYGON_TO_LIST (i->layer, polygon))
+    longjmp (i->env, 1);
+
+  return 0;
+}
+
+static int
+LOCtoLinePourPolygon_callback (const BoxType * b, void *cl)
+{
+  PourTypePtr pour = (PourTypePtr) b;
+  struct lo_info *i = (struct lo_info *) cl;
+
+  return r_search (pour->polygon_tree, (BoxType *) &i->polygon,
+                   NULL, LOCtoLinePolygon_callback, i);
+}
+
+static int
 LOCtoLinePad_callback (const BoxType * b, void *cl)
 {
   PadTypePtr pad = (PadTypePtr) b;
@@ -1958,19 +1982,11 @@ LookupLOConnectionsToLine (LineTypePtr Line, Cardinal LayerGroup,
           /* now check all polygons */
           if (PolysTo)
             {
-              printf ("Slow pour path for lines\n");
-              POUR_LOOP (LAYER_PTR (layer));
-              {
-                POURPOLYGON_LOOP (pour);
-                {
-                  if (!TEST_FLAG (TheFlag, polygon) &&
-                      IsLineInPolygon (Line, polygon) &&
-                      ADD_POLYGON_TO_LIST (layer, polygon))
-                    return True;
-                }
-                END_LOOP;
-              }
-              END_LOOP;
+              if (setjmp (info.env) == 0)
+                r_search (LAYER_PTR (layer)->pour_tree, (BoxType *) & info.line,
+                          NULL, LOCtoLinePourPolygon_callback, &info);
+              else
+                return True;
             }
         }
       else
@@ -2446,7 +2462,7 @@ static int
 LOCtoPolyPourPolygon_callback (const BoxType * b, void *cl)
 {
   PourTypePtr pour = (PourTypePtr) b;
-  struct lo_info *i = (struct pv_info *) cl;
+  struct lo_info *i = (struct lo_info *) cl;
 
   return r_search (pour->polygon_tree, (BoxType *) &i->polygon,
                    NULL, LOCtoPolyPolygon_callback, i);
